packages/databyss-services/pages/PageProvider.js

```diff
@@ -13,18 +13,30 @@ export const PageContext = createContext()
 const PageProvider = ({ children, initialState }) => {
   const [state, dispatch, stateRef] = useReducer(reducer, initialState)
 
-  const fetchPages = () => {
-    if (state.pages) {
-      return state.pages
+  const getPages = () => {
+    if (state.headerCache) {
+      return state.headerCache
     }
-    if (!state.isPagesLoading) {
-      dispatch(getPages())
+    if (!state.isLoading) {
```

Comment:
if this property is removed. Editor gets in a loop of fetching pages

Reply:
Ah, true.
So right now our cache values have 2 states based on type: `null` (unfetched) and `Page` (fetched)
What if we have 3 states: `null` (unfetched), `Promise` (fetching) and `Page` (fetched)

Note more explicit naming of pages -> pageHeaders throughout, and normalizing actions to `FETCH_` and `CACHE_` (actions should be command-like).

```js
// PageProvider.js
  const getPageHeaders = () => {
    if (state.headerCache) {
      return state.headerCache
    }
    if (!state.headerCache instanceof Promise) {
      dispatch(fetchPageHeaders())
    }
    return null
  }
// ...
export const PagesLoader = ({ children }) => {
  const { getPageHeaders } = usePageContext()
  const pages = getPageHeaders()

  if (pages instanceof Error) {
    return <ErrorFallback error={pages} />
  }

  if (typeof children !== 'function') {
    throw new Error('Child must be a function')
  }

  if (!pages || pages instanceof Promise) {
    return <Loading />
  }

  return children(pages)
}

// actions.js
export function fetchPageHeaders() {
  return dispatch => {
    const promise = services
      .getAllPages()
      .then(res => {
        dispatch({
          type: CACHE_PAGE_HEADERS,
          payload: res,
        })
      })
      .catch(e => {
        dispatch({
          type: CACHE_PAGE_HEADERS,
          payload: e,
        })
      })
    dispatch({
      type: FETCH_PAGE_HEADERS,
      payload: { promise },
    })
  }
}

// reducer.js
    case FETCH_PAGE_HEADERS: {
      return {
        ...state,
        headerCache: action.payload.promise,
      }
    }
```